### PR TITLE
Add nRF52840dk

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request adds the nrf52840 board, and tries to reuse as much of the nrf52dk board as possible.

Directory structure:

```
boards/
    nordic/
        nrf51dk          <--- nrf51dk board
        nrf52dk_base     <--- contains the shared setup code
        nrf52dk          <--- original dk board
        nrf52840dk       <--- new addition
```

### Testing Strategy

This pull request was tested by running apps on the nrf52840 board.

### TODO or Help Wanted

Maybe this isn't the right way to do this? To me this balances minimizing code copy (and overhead to update boards if things change) while supporting flexibility (e.g. the linker script has to change).

### Documentation Updated

- [x] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [x] Userland: Added/updated the application README, if needed.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
Do you imagine that nordic_base would also support the nrf51?

**Comment by niklasad1** [comment]
@bradjc Sorry for late replay but yes at least as long as we maintain `nrf51/dk`. Also, the registers are backward compatible so it should be pretty easy to add new chips/boards in the nrf5x family at least!

**Comment by bradjc** [comment]
@niklasad1 ok what do you think of what is there now. I made a nordic folder in boards, and moved the nrf51dk into it. I don't see any reason to split out the nrf51dk board file code into another file since it isn't shared with anything, so I left nrf52dk_base.

**Comment by niklasad1** [comment]
@bradjc I like it but I will probably not review it in detail before next week. I'm off to RustFest

**Comment by alevy** [comment]
Assigned @niklasad1 so he doesn't forget about it. I think this is sufficiently innocuous and obviously good, that as long as the organization is agreeable to folks doing NRF development, we should merge it.

In other words, I'm tacitly adding my approve pending @niklasad1's approve.

File: boards/nordic/nrf52840dk/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+# Upload the kernel over JTAG
+.PHONY: flash
+flash: target/$(TARGET)/release/$(PLATFORM).hex
+	$(JLINK) $(JLINK_OPTIONS) $(JLINK_SCRIPTS_DIR)/flash-kernel.jlink
+
+# Upload the kernel over serial/bootloader
+.PHONY: program
+program: target/$(TARGET)/release/$(PLATFORM).hex
+	$(error Cannot program nRF52DK over USB. Use \`make flash\` and JTAG)
```
Comment (niklasad1) [inline boards/nordic/nrf52840dk/Makefile]: Can we change `nRF52DK` here to `$(PLATFORM)`?

File: boards/nordic/nrf52840dk/src/io.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        Ok(())
+    }
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+#[lang = "panic_fmt"]
+/// Panic handler
+pub unsafe extern "C" fn panic_fmt(args: Arguments, file: &'static str, line: u32) -> ! {
+    // The nRF52 DK LEDs (see back of board)
```
Comment (niklasad1) [inline boards/nordic/nrf52840dk/src/io.rs]: remove the comment or rename nRF52 to nRF52840 to not confuse people!

File: boards/nordic/nrf52840dk/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+extern crate capsules;
+#[allow(unused_imports)]
+#[macro_use(debug, debug_verbose, debug_gpio, static_init)]
+extern crate kernel;
+extern crate nrf52;
+extern crate nrf52dk_base;
+extern crate nrf5x;
+
+// The nRF52 DK LEDs (see back of board)
```
Comment (niklasad1) [inline boards/nordic/nrf52840dk/src/main.rs]: remove the comment or rename nRF52 to nRF52840 to not confuse people!

File: boards/nordic/nrf52840dk/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+const LED1_PIN: usize = 13;
+const LED2_PIN: usize = 14;
+const LED3_PIN: usize = 15;
+const LED4_PIN: usize = 16;
+
+// The nRF52 DK buttons (see back of board)
```
Comment (niklasad1) [inline boards/nordic/nrf52840dk/src/main.rs]: remove the comment or rename nRF52 to nRF52840 to not confuse people!

**Pull request merged**